### ISSUE TITLE
SELinux prevents omsconfig logrotate on CentOS

### ISSUE THREAD

**Issue by kule** [Microsoft/OMS-Agent-for-Linux#781]
Similar to https://github.com/Microsoft/OMS-Agent-for-Linux/issues/172 this stops logrotate on CentOS.

SELinux logs the following in /var/log/audit/audit.log:

    type=AVC msg=audit(1536723902.222:511566): avc:  denied  { getattr } for  pid=50096 comm="logrotate" path="/var/opt/microsoft/omsconfig/omsconfig.log" dev="dm-5" ino=133401 scontext=system_u:system_r:logrotate_t:s0-s0:c0.c1023 tcontext=system_u:object_r:var_t:s0 tclass=file

    type=AVC msg=audit(1536723902.222:511567): avc:  denied  { getattr } for  pid=50096 comm="logrotate" path="/var/opt/microsoft/omsconfig/omsconfigdetailed.log" dev="dm-5" ino=133498 scontext=system_u:system_r:logrotate_t:s0-s0:c0.c1023 tcontext=system_u:object_r:var_t:s0 tclass=file

It seems to me like the two omsconfig logs (````omsconfig.log```` & ````omsconfigdetailed.log````) could do with being put into their own log directory so it could be labelled separately then this issue could be fixed with the following commands (run as root):

semanage fcontext -a -t var_log_t '/var/opt/microsoft/omsconfig/log(/.*)?'
restorecon -R -v /var/opt/microsoft/omsconfig/log

**Comment by dtonhofer** [comment]
Yes, observing this on **CentOS 7.6** and **omsconfig-1.1.1-831.x86_64**

**Comment by scottolsenbp** [comment]
For the time being, this *appears* to suffice:

semanage fcontext -a -t var_log_t '/var/opt/microsoft/omsconfig(/.*.log)?'
restorecon -R -v /var/opt/microsoft/omsconfig

**Comment by judyanndixon** [comment]
I was seeing this in the audit.log.
type=AVC msg=audit(1569585001.586:996528): avc:  denied  { getattr } for  pid=114336 comm="logrotate" path="/var/opt/microsoft/omsagent/LAD/log/omsagent.log" dev="dm-4" ino=108 scontext=system_u:system_r:logrotate_t:s0-s0:c0.c1023 tcontext=system_u:object_r:var_t:s0 tclass=file permissive=0

I fixed it doing this.
semanage fcontext -a -t logrotate_exec_t '/var/opt/microsoft/omsagent/LAD/log/omsagent.log'
restorecon -v '/var/opt/microsoft/omsagent/LAD/log/omsagent.log'

Now I am seeing these.
type=AVC msg=audit(1569573902.078:944320): avc:  denied  { getattr } for  pid=60870 comm="logrotate" path="/var/opt/microsoft/omsagent/bb50a759-f3c9-40be-9582-24221564a97f/log/omsagent.log" dev="dm-4" ino=117 scontext=system_u:system_r:logrotate_t:s0-s0:c0.c1023 tcontext=system_u:object_r:var_t:s0 tclass=file permissive=0

**Comment by scottolsenbp** [comment]
FYI, I believe that the **var_log_t** label is unfortunately being applied to the directory /var/opt/microsoft/omsconfig directory _after_ the initial creation of these two log files.  

I've checked this morning after applying to the **var_log_t** label to the two existing log files last evening and logrotate was successful in rotating them this time.  The newly created gzip archive file also inherited the proper **var_log_t** label as well.

**Comment by kamsalisbury** [comment]
Seeing this issue on;
CentOS Linux release 7.7.1908 (Core)
and
omsconfig-1.1.1-926.x86_64

**Comment by scottolsenbp** [comment]
This appears to have popped up elsewhere again;

CentOS Linux release 7.7.1908 (Core)
omsagent-1.12.15-0

Mar 20 03:49:03 XXXXXXXXXXXXXXXXX python: SELinux is preventing /usr/sbin/logrotate from getattr access on the file /var/opt/microsoft/omsagent/XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX/log/omsagent.log.#012#012*****  Plugin catchall_labels (83.8 confidence) suggests   *******************#012#012If you want to allow logrotate to have getattr access on the omsagent.log file#012Then you need to change the label on /var/opt/microsoft/omsagen...

**Comment by nateoconnell** [comment]
Commenting to show additional interest in having the omsconfg and omsconfigdetailed log files moved to /var/opt/microsoft/omsconfig/log where the var_log_t type can be applied. Currently running into disk space utilization issues on /var on about 81 servers partly due to the omsconfig logs not rotating with SELinux set to enforcing.
```
  38 Red Hat Enterprise Linux Server release 7.6 (Maipo) omsagent-1.12.25-0.x86_64
  32 Red Hat Enterprise Linux Server release 7.4 (Maipo) omsagent-1.12.25-0.x86_64
   8 Red Hat Enterprise Linux Server release 7.5 (Maipo) omsagent-1.12.25-0.x86_64
   1 Red Hat Enterprise Linux Server release 7.8 (Maipo) omsagent-1.12.25-0.x86_64
   1 Red Hat Enterprise Linux Server release 7.6 (Maipo) omsagent-1.12.15-0.x86_64
   1 CentOS Linux release 7.7.1908 (Core) omsagent-1.12.25-0.x86_64
```
Several workarounds have been considered, but the best approach would seem to be for the logs to be in their own sub-directory as the initial comment suggests.

By default new file/directory objects will inherit the SELinux type of their parent directory, even if there are fcontext rules that specify a different type, like those below. Setting the type of /var/opt/microsoft/omsconfig to var_log_t, would cause any new objects created under it by the omsagent to inherit that type. With the rules below, a `restorecon -rv /var/opt/microsoft/omsconfig` ensures that any files/directories that do not have '.log' in the name, like the run/ and status/ directories, have the var_t type that /var/opt/microsoft/omsconfig ships with. This does not help for new files or directories that the omsagent might create.
```
# semanage fcontext -a -t var_log_t '/var/opt/microsoft/omsconfig(/.*\.log(-[0-9]{8}(.gz)?)?)?'
# semanage fcontext -a -t var_t '/var/opt/microsoft/omsconfig/(?!.*?.log).*'

# semanage fcontext -lC
[. . .]
/var/opt/microsoft/omsconfig(/.*\.log(-[0-9]{8}(.gz)?)?)? all files          system_u:object_r:var_log_t:s0
/var/opt/microsoft/omsconfig/(?!.*?.log).*         all files          system_u:object_r:var_t:s0
```
One way to ensure that newly created files have the type configured in the fcontext rules would be for the administrator or application creating the file or directory to run a restorecon on it right after creation. There is a restorecond daemon that can be installed and set up to watch certain directories for creation of new files/directories and apply the types configured in the fcontext rules to those new objects. It is acknowledged that this can lead to race conditions. If the omsagent creates a file and attempts to access it before the correct context is applied by restorecond then it will be blocked by SELinux, so this may not be an ideal approach.

It is also possible to compile and install a custom policy module that will result in creation of specific named files with the correct context using the File Name Transition feature of SELinux [1]. One limitation here is that the there cannot be any globbing or regular expressions in the file names that are specified in the policy. Currently, we need the /var/opt/microsoft/omsconfig directory to have the var_log_t so that logrotate is allowed to rotate the log files in it, so every file or directory that might be created under /var/opt/microsoft/omsconfig/ that needs to be created with the var_t rather than var_log_t would need to be specified explicitly. This approach does not seem to be sustainable, unless it is known that, besides the log files, only the run/ and status/ directories or some other limited set of non-log files are expected to be created there.

Another workaround, rather than adding fcontext rules and applying the var_log_t type to /var/opt/microsoft/omsconfig and the log files with restorecon, would be to set the logrotate_t domain to permissive mode [2], leaving the rest of the system in enforcing mode. This would mean that logrotate would be allowed to violate SELinux policy rules and that these events would only be logged, not denied. The omsconfig logs could then be rotated without having to fully disable SELinux on the systems where it is running in enforcing mode.

But the best option seems to be for the omsagent to log to a sub-directory with the right type.

[1]
https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/selinux_users_and_administrators_guide/sec-file-name-transition
> Note that file name transition uses an exact match done by the strcmp() function. Use of regular expressions or wildcard characters is not considered.

https://danwalsh.livejournal.com/73796.html

[2]
To set logrotate domain to permissive:
`# semanage permissive -a logrotate_t`
To verify:
`# semanage permissive -l`

**Comment by silverl** [comment]
FYI, this remains an issue on `centos-release-7-8.2003.0.el7.centos.x86_64` and `omsconfig-1.1.1-930.x86_64`

**Comment by mikenri** [comment]
I am also seeing this issue on Red Hat Enterprise Linux 7.9

**Comment by Klaas-** [comment]
I think some of you may be hitting something that was recently fixed, https://github.com/microsoft/omi/issues/680 and https://github.com/microsoft/SCXcore/issues/168 , basically an update used to unload the custom selinux modules

**Comment by Klaas-** [comment]
So Microsoft came back to me and said they'll never fix it, I am guessing we can consider this project abandoned ....
